tests/index.js

```diff
@@ -2,116 +2,248 @@ const test = require('tape');
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+        },
+        'Can parse URI with payload starting with `0x`'
+    );
+
+    t.deepEqual(
+        parse('ethereum:pay-0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD'),
+        {
+            scheme: 'ethereum',
+            prefix: 'pay',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+        },
+        'Can parse URI with payload starting with `0x` and `pay` prefix'
+    );
+
+    t.deepEqual(
+        parse('ethereum:foo-0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD'),
+        {
+            scheme: 'ethereum',
+            prefix: 'foo',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+        },
+        'Can parse URI with payload starting with `0x` and `foo` prefix'
+    );
+
+    t.deepEqual(
+        parse('ethereum:foo-doge-to-the-moon.eth'),
+        {
+            scheme: 'ethereum',
+            prefix: 'foo',
+            target_address: 'doge-to-the-moon.eth',
+        },
+        'Can parse URI with an ENS name'
+    );
+
+    t.deepEqual(
+        parse('ethereum:0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD@42'),
+        {
+            scheme: 'ethereum',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+            chain_id: '42',
+        },
+        'Can parse URI with chain id'
+    );
+
+    t.deepEqual(
+        parse(
+            'ethereum:0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD/transfer?address=0x12345&uint256=1'
+        ),
+        {
+            scheme: 'ethereum',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+            function_name: 'transfer',
+            args: [
+                ['address', '0x12345'],
+                ['uint256', '1'],
+            ],
+        },
+        'Can parse an ERC20 token transfer'
+    );
+
+    t.deepEqual(
+        parse(
+            'ethereum:0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD?value=2.014e18&gas=10&gasLimit=21000&gasPrice=50'
+        ),
+        {
+            scheme: 'ethereum',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+            parameters: {
+                value: '2014000000000000000',
+                gas: '10',
+                gasLimit: '21000',
+                gasPrice: '50',
+            },
+        },
+        'Can parse a url with value and gas parameters'
+    );
+
+    t.deepEqual(
+        parse(
+            'ethereum:0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD?value=-2.014e18&gas=10&gasLimit=21000&gasPrice=50'
+        ),
+        {
+            scheme: 'ethereum',
+            target_address: '0x1234DEADBEEF5678ABCD1234DEADBEEF5678ABCD',
+            parameters: {
+                value: '-2014000000000000000',
+                gas: '10',
+                gasLimit: '21000',
+                gasPrice: '50',
+            },
+        },
+        'Can parse a url with negative value and gas parameters'
```

Comment:
can you explain this case? AFAIK this shouldn't be valid

Reply:
You're right, it shouldn't take negative value for `value`, `gas`, `gasLimit`, and `gasPrice`, but it **should** allow for negative values for arbitrary arguments.

Reply:
Do you think pertaining to arguments we should filter non-solidity-types or just allow any arbitrary string value to be passed? Personally I think arbitrary values would allow for support for further additions to the spec without updates but we could always have it filter stuff.

Reply:
I will adjust the testing accordingly